README.md
# RDatastream

RDatastream is a R interface to the [Thomson Dataworks Enterprise](http://dataworks.thomson.com/Dataworks/Enterprise/1.0/) SOAP API (non free), with some convenience functions for retrieving Datastream data specifically. This package requires valid credentials for this API.

## Notes

* This API happens to be the one used by the [MATLAB datafeed toolbox](http://www.mathworks.fr/help/toolbox/datafeed/datastream.html), so if you have used it before, this package should work as well.
* This package is mainly meant to access Datastream. It should work for [other Dataworks Enterprise sources](http://dtg.tfn.com/data/), but they are quite poorly documented and I do not have valid credentials for most. If you do, and want to see this package extended, please get in touch!
* Not using `SSOAP` is deliberate; I initially toyed with it, and it felt too cumbersome for such a simple API.
* If you feel like some of design choices could be improved, please also get in touch!

## Installation

First, you will need dependencies `XML` and `RCurl`.

    install.packages("XML")
    install.packages("RCurl")

For now, the easiest way to install RDatastream is to use the `devtools` package to get the latest version straight from Github. Install the `devtools` package if you do not have it yet:

    install.packages("devtools")
    
Then install `RDatastream` from Github.

    devtools::install_github("fcocquemas/RDataStream")

## Basic use

First, you need to define a user with your valid credentials, like this:

    user <- list(username = "DS:XXXX000", password = "XXX000")

Then you can load the library check which sources you have access to with these credentials:

    library(RDatastream)
    dsSources(user)

Hopefully "Datastream" should be among the sources.

Simple requests can then be made. Let's say, for instance, that we want the price and market value of IBM (quoted on the NYSE) on June 4th, 2007. The NYSE tickers are preceded by `"U:"`, so the DS ticker is `"U:IBM"`.

    dat <- ds(user, securities = "U:IBM", fields = c("P", "MV"), date = "2007-06-04")

Or equivalently:

    dat <- ds(user, "U:IBM", c("P", "MV"), "2007-06-04")
    
Checking `data` should show the status code and, if need be, the error message. To look at the data returned as a dataframe, do:

    dat[["Data",1]]
    
Which should be:

      CCY       DATE                DISPNAME FREQUENCY       MV      P SYMBOL
    1  U$ 2007-06-04 INTERNATIONAL BUS.MCHS.         D 157733.1 106.23  U:IBM

You can also specify several tickers and date ranges instead of a single date. For instance, let's add Microsoft (`"@MSFT"`, NASDAQ tickers are preceded by `"@"`), and let's look from June 4th, 2007 to June 4th, 2009 at the monthly frequency.

    dat <- ds(user, c("U:IBM", "@MSFT"), c("P", "MV"), 
              fromDate = "2007-06-04", toDate = "2009-06-04", period = "M")

As you can seen, each ticker is dealt with in a separate record. To get access to the resulting dataframes, just do:

    dat["Data",]

## Advanced use

### Using custom requests

The Datastream request syntax is somewhat arcane but can be more powerful in certain cases. A [decent guide can be found here](http://dtg.tfn.com/data/DataStream.html). You can use this syntax directly with this package when your needs are more sophisticated.

For instance, let's say I want the data from the previous example combined in a single dataframe.

    request1 <- "U:IBM,@MSFT~=P,MV~2007-06-04~:2009-06-04~M"
    dat <- ds(user, requests = request1)
    dat[["Data",1]]
    
We can run several such requests in a single API call.

    request2 <- "U:MMM~=P,PO~2007-09-01~:2007-09-12~D"
    request3 <- "906187~2008-01-01~:2008-10-02~M"
    request4 <- "PCH#(U:BAC(MV))~2008-01-01~:2008-10-02~M"
    requests <- c(request1, request2, request3, request4)
    dat <- ds(user, requests = requests)
    dat["Data",]

### Other useful tips with the Datastream syntax

#### Get some reference information on a security with `"~XREF"`, including ISIN, industry, etc.

    dat <- ds(user, requests = "U:IBM~XREF") 
    dat[["Data",1]]
    
#### Get some static items like NAME, ISIN with `"~REP"`

    dat <- ds(user, requests = "U:IBM~=NAME,ISIN~REP") 
    dat[["Data",1]]

#### Convert the currency e.g. to Euro with `"~~EUR"`

    dat <- ds(user, requests = "U:IBM(P)~~EUR~2007-09-01~:2009-09-01~D") 
    dat[["Data",1]]
    
#### Use Datastream expressions, e.g. for a moving average on 20 days

    dat <- ds(user, requests = "MAV#(U:IBM,20D)~2007-09-01~:2009-09-01~D") 
    dat[["Data",1]]

User-defined expressions require using a syntax similar to: `E237(U:IBM)~2010-01-01~:2015-01-01~D~#USERNAME` (thanks @abbanerjeersm).

Any other tip we should know about?

## Resources

It is recommended that you read the [Thomson Dataworks Enterprise User Guide](http://dataworks.thomson.com/Dataworks/Enterprise/1.0/documentation/user%20guide.pdf), especially section 4.1.2 on client design. It gives reasonable guidelines for not overloading the servers with too intensive requests.

For building custom Datastream requests, useful guidelines are given on this somewhat old [Thomson Financial Network](http://dtg.tfn.com/data/DataStream.html) webpage. I have been able to replicate all of their examples except for the Navigator search ones.

If you have access codes for the Datastream Extranet, you can use the [Datastream Navigator](http://product.datastream.com/navigator/) to look up codes and data types.

## Licence

RDatastream is released under the MIT licence.


CHANGELOG
21/04/2017
==========
- BUGFIX: Do not assume the fields are numeric (introduced NA in some cases). Thanks to Matthias Leute for reporting the issue.

04/06/2013
==========
- BUGFIX: Missing currency (e.g. price index field) was causing a bug in the
  conversion to data.frame. All NULL fields are now changed to NA. Thanks to 
  Claudio Ortelli for finding the issue.

R/sources.R
#' Find available sources from Thomson Reuters Datastream SOAP API.
#' 
#' @export
#' @param user list with values username and password.
#' @return matrix with the available sources.
#' @examples
#' user <- list(username= "DS:XXXX000", password="XXX000")
#' dsSources(user)
dsSources <- function(user) {
  wsdl <- "http://dataworks.thomson.com/Dataworks/Enterprise/1.0/webServiceClient.asmx"
  xmlns <- "http://xml.thomson.com/financial/v1/tools/distribution/dataworks/enterprise/2003-07/"
 
  
  # SOAP headers
  xmlRequest <- '<?xml version="1.0" encoding="utf-8"?>
<soap12:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap12="http://www.w3.org/2003/05/soap-envelope">
  <soap12:Body>'
  
  # SOAP action
  xmlRequest <- paste(xmlRequest, '<Sources xmlns="',
                      xmlns, '">', sep="")
  
  # Add user
  xmlRequest <- paste(xmlRequest, '<User>',
                      '<Username>', user$username, '</Username>',
                      '<Password>', user$password, 
                      '</Password></User>', sep="")
  
  # SOAP footer
  xmlRequest <- paste(xmlRequest, '</Sources>',
                      ' </soap12:Body></soap12:Envelope>', sep="")
  
  # Perform the request
  resp <- getURL(wsdl, 
                 httpheader=c(Accept="text/xml", Accept="multipart/*", 
                              'Content-Type' = "application/soap+xml; charset=utf-8"),
                 postfields=xmlRequest, verbose = FALSE)
  resp <- xmlParse(resp)
  
  # Extract the XML body and convert it to list
  sources <- xmlChildren(resp[["/soap:Envelope/soap:Body/*"]][[1]])
  sources <- sapply(sources, xmlToList)
  sources
}


R/ds.R
#' Request data from Thomson Reuters Datastream SOAP API.
#' 
#' @export
#' 
#' @param user list with values username and password.
#' @param securities vector of DS tickers or codes.
#' @param fields vector of DS datatypes.
#' @param date date in yyyy-mm-dd format, if only one day is requested.
#'   If set, ignores fromDate, toDate and period.
#' @param fromDate date in yyyy-mm-dd format from which to start series.
#' @param toDate date in yyyy-mm-dd format at which to end series.
#' @param period character describing the periodicity ('D','W','M','Q','Y'),
#'   defaults to daily.
#' @param requests a vector of requests following the Datastream format.
#'   If set, ignores all other parameters (except user). More flexible
#'   syntax, notably for the use of expressions.
#' @param asDataFrame boolean whether to output a dataframe in the "Data"
#'   row of the returned matrix.
#' @param source default "Datastream", useful if you want to access another
#'   Dataworks Enterprise data source. You can obtain the list of sources
#'   you have access to by using the dsSources() function in this package.
#' @return matrix with the returned data, columns being individual requests
#'   and rows "Source", "Instrument", "StatusType", "StatusCode", 
#'   "StatusMessage", "Fields" and "Data".
#' @examples
#' user <- list(username= "DS:XXXX000", password="XXX000")
#' ds(user, c("U:IBM", "U:MMM"), c("P", "PO"), "2012-07-20")
#' 
#' request1 <- "U:IBM(P)~~USD~2007-09-01~:2008-09-05~D"
#' request2 <- "U:BAC~=P,PO~2007-09-01~:2007-09-12~D"
#' request3 <- "906187~2008-01-01~:2008-10-02~M"
#' request4 <- "PCH#(U:MMM(MV))~2008-01-01~:2008-10-02~M"
#' requests <- c(request1, request2, request3, request4)
#' 
#' ds(user, requests=request1)
#' ds(user, requests=requests)
ds <- function(user, securities=NULL, fields=NULL, 
               date=NULL, fromDate=NULL, toDate=NULL,
               period="D", requests=NULL, asDataFrame=TRUE,
               source="Datastream") {
  wsdl <- "http://dataworks.thomson.com/Dataworks/Enterprise/1.0/webServiceClient.asmx"
  xmlns <- "http://xml.thomson.com/financial/v1/tools/distribution/dataworks/enterprise/2003-07/"
  
  # If requests is not set, create it from the pieces
  if(is.null(requests)) {
    requests <- securities
    if(!is.null(fields)) {
      requests <- paste(requests, "~=", paste(fields, collapse=","), sep="")
    }
    
    if(!is.null(date)) {
      requests <- paste(requests, "~@", date, sep="")
    } else {
      if(!is.null(fromDate)) {
        requests <- paste(requests, "~", fromDate, sep="")
      }
      if(!is.null(toDate)) { 
        requests <- paste(requests, "~:", toDate, sep="")
      }
      requests <- paste(requests, "~", period, sep="")
    }
  }
  
  # SOAP headers
  xmlRequest <- '<?xml version="1.0" encoding="utf-8"?>
<soap12:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soap12="http://www.w3.org/2003/05/soap-envelope">
  <soap12:Body>'
  
  # SOAP action
  xmlRequest <- paste(xmlRequest, '<RequestRecordsAsXml xmlns="',
                      xmlns, '">', sep="")
  
  # Add user
  xmlRequest <- paste(xmlRequest, '<User>',
                      '<Username>', user$username, '</Username>',
                      '<Password>', user$password, 
                      '</Password></User><Requests>', sep="")
  
  # Add the requests 
  xmlRequest <- paste(xmlRequest, 
                      paste('<RequestData><Source>', source, '</Source><Instrument>',
                            requests, '</Instrument>',
#                             ifelse(is.null(options), '', 
#                                    paste('<Options>', options, '</Options>', sep="")),
                            '</RequestData>', 
                            sep="", collapse=""), sep="")
  
  # SOAP footer
  xmlRequest <- paste(xmlRequest, '</Requests>',
                      '<RequestFlags>0</RequestFlags>',
                      '</RequestRecordsAsXml>',
                      ' </soap12:Body></soap12:Envelope>', sep="")
  
  # Perform the request
  resp <- getURL(wsdl, 
                 httpheader=c(Accept="text/xml", Accept="multipart/*", 
                              'Content-Type' = "application/soap+xml; charset=utf-8"),
                 postfields=xmlRequest, verbose = FALSE)
  resp <- xmlParse(resp)
  
  # Extract the XML body and convert it to list
  records <- xmlChildren(resp[["/soap:Envelope/soap:Body/*"]][[1]][[1]])
  records <- sapply(records, xmlToList)
  
  # Convert the obtained data to a dataframe
  if(asDataFrame) {
    records <- rbind(records, apply(records, 2, function(record) {
      as.data.frame(sapply(names(record$Fields), function(fieldname){
        field <- record$Fields[[fieldname]]
        
        # Fix some NULL fields when e.g. no currency (price index variable)
        if(is.null(field)) field <- NA
        
        if(fieldname == "DATE") {
          list(as.Date(unlist(field)))
        } 
        else if(is.list(field)) {
          list(type.convert(gsub("NaN", NA, as.character(unlist(field)))))
        }
        else {
          list(rep(field, 1, length(record$Fields$DATE)))
        }
      }))
    }))
    rownames(records)[7] <- "Data"
  }
  
  records
}

